runners/flink/src/test/java/org/apache/beam/runners/flink/PortableExecutionTest.java

```diff
@@ -0,0 +1,234 @@
+                        controlServer.getApiServiceDescriptor(),
+                        ManagedChannelFactory.createDefault(),
+                        //InProcessManagedChannelFactory.create(),
+                        HarnessStreamObserverFactories.fromOptions(
+                            PipelineOptionsFactory.create()));
+                  } catch (Exception e) {
+                    throw new RuntimeException(e);
+                  }
+                });
+        try {
+          InstructionRequestHandler controlClient =
+              clientSource.take(workerId, Duration.ofSeconds(2));
+          return (container) -> RemoteEnvironment.forHandler(container, controlClient);
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      }
+    };
+  }
+
+  @Before
+  public void setup() {
+    DockerJobBundleFactory.FACTORY.set(this::createJobBundleFactory);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    sdkHarnessExecutor.shutdownNow();
+    try {
+      sdkHarnessExecutorFuture.get();
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof RuntimeException
+          && e.getCause().getCause() instanceof InterruptedException) {
+        // expected
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  private static ArrayList<KV<String, Iterable<Long>>> outputValues = new ArrayList<>();
+
+  @Test
+  public void testExecution() throws Exception {
+    Pipeline p = Pipeline.create();
+    p.apply("impulse", Impulse.create())
+        .apply(
+            "create",
+            ParDo.of(
+                new DoFn<byte[], String>() {
+                  @ProcessElement
+                  public void process(ProcessContext ctxt) {
+                    ctxt.output("zero");
+                    ctxt.output("one");
+                    ctxt.output("two");
+                  }
+                }))
+        .apply(
+            "len",
+            ParDo.of(
+                new DoFn<String, Long>() {
+                  @ProcessElement
+                  public void process(ProcessContext ctxt) {
+                    ctxt.output((long) ctxt.element().length());
+                  }
+                }))
+        .apply("addKeys", WithKeys.of("foo"))
+        // Use some unknown coders
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), BigEndianLongCoder.of()))
+        // Force the output to be materialized
+        .apply("gbk", GroupByKey.create())
+        .apply(
+            "collect",
+            ParDo.of(
+                new DoFn<KV<String, Iterable<Long>>, Void>() {
+                  @ProcessElement
+                  public void process(ProcessContext ctx) {
+                    outputValues.add(ctx.element());
+                  }
+                }));
+
+    RunnerApi.Pipeline pipelineProto = PipelineTranslation.toProto(p);
+
+    outputValues.clear();
+    // execute the pipeline
+    ListeningExecutorService executor =
```

Comment:
Set up this executor inside `@Before` and do the shutdown in `@After` so an exception that is thrown during the test still cleans up the executor.

Reply:
done